README.md
# myHome-Doorcam

Welcome to the myHome Doorcam project.

The Doorcam is a replacement for video doorbells that invade your privacy.

Using off-the-shelf hardware the Doorcam will do all of the usual stuff, such as capturing video on motion events, as well as more advanced stuff, like face-detection.  

Getting Started
---------------
To get started you will need a few things:


1. Clone the project 
````
mkdir $HOME/livefree & cd $HOME/livefree
git clone https://github.com/bluesunrise/myhome.git
````

2. [Raspberry Setup](./SETUP.md)

3. Running
````
    cd viewer
    python viewer-tk.py
````    

4. [DoorCam Viewer Usage](./viewer/README.md)

docs/history.txt
    1  ls
    2  hostname
    3  hostname -I
    4  ls -al
    5  java
    6  java -version
    7  python -version
    8  python --version
    9  java -version
   10  raspistill  -o Desktop/image.jpg
   11  raspistill  --help
   12  raspistill  --demo
   13  raspistill  --preview
   14  raspivid
   15  raspivid -t 0
   16  python --version
   17  python3
   18  which python
   19  which python3
   20  ls /usr/local/opt/python/libexec/bin
   21  ls /usr/local/opt/python/
   22  which python3
   23  ls -al /usr/bin/python3
   24  pip3
   25  pip
   26  pip3
   27  pip install virtualenv virtualenvwrapper
   28  pwd
   29  vi .bashrc
   30  source ~/.bashrc
   31  l .local
   32  l
   33  ll
   34  pwd
   35  vi .bashr
   36  vi .bashrc
   37  ls -CF
   38  ll
   39  ls .local/
   40  ls .local/bin
   41  vi .bashrc
   42  pwd
   43  vi .bashrc
   44  source .bashrc
   45  ls -al /usr/local/bin/
   46  virtualenv
   47  which virtualenvwrapper.sh
   48  vi .bashrc
   49  source .bashrc
   50  vi .bashrc
   51  source .bashrc
   52  /usr/bin/python3
   53  virtualenv
   54  virtualenvwrapper.sh 
   55  pip install virtualenv
   56  pip install virtualenvwrapper
   57  virtualenvwrapper.sh 
   58  pip3 install virtualenv
   59  pip3 install virtualenvwrapper
   60  virtualenvwrapper.sh
   61  l
   62  vi .bashrc
   63  alias l='ls -al'
   64  l
   65  ls -l
   66  vi .bashrc
   67  exit
   68  cd livefree/myhome/
   69  ls
   70  cd simpledemo/
   71  ls
   72  LD_PRELOAD=/usr/lib/arm-linux-gnueabihf/libatomic.so.1 python demo.py
   73  export LD_PRELOAD=/usr/lib/arm-linux-gnueabihf/libatomic.so.1
   74  python demo.py 
   75  DF
   76  df
   77  vi ~/.bashrc 
   78  exit
   79  ps aux 
   80  ps aux  | python
   81  ps 
   82  kill -9 15503
   83  ps aux  | grep python
   84  ls -lexit
   85  exit
   86  l
   87  workon
   88  mkvirtualenv myhome -p python3
   89  workon
   90  l /home/pi/.local/bin/
   91  cd .virtualenvs/
   92  l
   93  cd myhome
   94  ls
   95  ls bin
   96  ls -al
   97  workon myhome
   98  cd ..
   99  mkproject myhome -p python3
  100  df
  101  rm -rf myhome
  102  cd 
  103  python
  104  which mkvirtualenv
  105  mkvirtualenv
  106  mkvirtualenv cv2
  107  echo $WORKON_HOME
  108  echo $PATH
  109  l .virtualenvs/cv2/bin/activate
  110  .virtualenvs/cv2/bin/activate
  111  l .virtualenvs/cv2/bin
  112  vi .virtualenvs/cv2/bin/activate
  113  chmod +x .virtualenvs/cv2/bin/activate
  114  workon c2
  115  workon cv2
  116  source ~/.bashrc
  117  workon cv2
  118  vi .bashrc
  119  python
  120  python3
  121  date-alternatives --list python
  122  update-alternatives --list python
  123  update-alternatives
  124  update-alternatives --list python
  125  update-alternatives --list python3
  126  update-alternatives --install /usr/bin/python python /usr/bin/python3
  127  update-alternatives --install /usr/bin/python python /usr/bin/python3 1
  128  sudo update-alternatives --install /usr/bin/python python /usr/bin/python3 1
  129  update-alternatives --list python
  130  python
  131  workon cv
  132  mkvirtualenv cv
  133  pip
  134  pip --version
  135  apt list --installed
  136  apt list --installed | grep libhdf5
  137  apt list python*opencv*
  138  apt show python3-opencv
  139  pwd
  140  mkdir livefree
  141  cd livefree
  142  mkdir myhome
  143  git
  144  cd ..
  145  rmdir myhome
  146  ls
  147  pwd
  148  cd livefree
  149  l
  150  rmdir myhome
  151  git clone https://github.com/bluesunrise/myhome.git
  152  ls
  153  cd myhome
  154  ls
  155  cd simpledemo/
  156  l
  157  more demo.py 
  158  python demo.py 
  159  pip list opencv-python
  160  pip
  161  pip search opencv-python
  162  pip install opencv-python
  163  ls -al
  164  python demo.py 
  165  sudo apt-get update && sudo apt-get upgrade
  166  df 
  167  sudo apt-get install libjpeg-dev libtiff5-dev libjasper-dev libpng-dev
  168  sudo apt-get install build-essential cmake pkg-config
  169  sudo apt-get install libavcodec-dev libavformat-dev libswscale-dev libv4l-dev
  170  sudo apt-get install libxvidcore-dev libx264-dev
  171  sudo apt-get install libgdk-pixbuf2.0-dev libpango1.0-dev
  172  sudo apt-get install libgtk2.0-dev libgtk-3-dev
  173  sudo apt-get install libhdf5-dev libhdf5-serial-dev libhdf5-103
  174  sudo apt-get install libqtgui4 libqtwebkit4 libqt4-test python3-pyqt5
  175  python demo.py 
  176  udo apt-get install libcblas-dev
  177  sudo apt-get install libhdf5-dev
  178  sudo apt-get install libhdf5-serial-dev
  179  sudo apt-get install libatlas-base-dev
  180  python demo.py 
  181  sudo apt-get install libcblas-dev
  182  pip uninstall opencv-python
  183  pip uninstall opencv-contrib-python
  184  pip install opencv-contrib-python
  185  saaaaaaaaaaaaaaaaaaaaa
  186  python demo.py 
  187  pip3 install opencv-python; sudo apt-get install -y libcblas-dev libhdf5-dev libhdf5-serial-dev libatlas-base-dev libjasper-dev  libqtgui4  libqt4-test
  188  sudo apt-get install libatlas-base-dev
  189  python demo.py
  190  LD_PRELOAD=/usr/lib/arm-linux-gnueabihf/libatomic.so.1 python demo.py 
  191  history
  192  pwd
  193  history > ~/history.txt
  194  more ~/history.txt 
  195  cd ../
  196  l
  197  cd facedemo/
  198  vi facestreamdemo.py 
  199  cd livefree/myhoe
  200  cd livefree/myhome
  201  ls
  202  cd facedemo/
  203  ls
  204  python facedemo.py 
  205  vi facestreamdemo.py 
  206  python facestreamdemo.py 
  207  cd livefree/myhome/
  208  git status
  209  git pull
  210  cd ..
  211   python3 ./opencv-face-recognition/recognize_video.py -s ./facedemo/fools-in-california.mp4
  212  pip --version
  213  pip install imutils
  214  pip install sklearn
  215  ls
  216  vi zoom.py
  217  ls -al
  218  vi zoom.py
  219  mkdir /usr/local/groundwork/config2
  220  ls
  221  cd livefree
  222  ls
  223  cd myhome
  224  git status
  225  git diff
  226  git status
  227  git checkout -- facedemo/facestreamdemo.py
  228  git pull
  229  history
  230  python

viewer/viewer-tk.py

import argparse
import functools
import glob
import json
import logging
import os
import pickle
import queue
import threading
import time
from collections import namedtuple
from datetime import datetime, timedelta

import cv2
import imutils
import numpy as np

from common.bvstream import BufferedVideoStream
from common.card import Card, cardsOverlap
from common.clipmaker import ClipMaker
from common.facerecognizer import FaceRecognizer
from common.facerecorder import FaceRecorder
from common.fd import extractEmbeddings, trainModel
from common.tkapp import TkApp
from motion.motiondetection import initializeMotionDetection

MDState = namedtuple("MDState", "detected trackedFrame trackedAt lastDetectedAt")


logger = logging.getLogger(__name__)
tkApp = None
initializeMotionDetection()

def getCLIArgs():
    ap = argparse.ArgumentParser()
    ap.add_argument("--action", type=int, default="1", choices=[0, 1, 2],
                    help="action level 0:motion detection, 1:face detection, 2:face recognition")
    ap.add_argument("--face-detector", type=str,
                    default=relPath(["models"]),
                    help="path to OpenCV's deep learning face detector")
    ap.add_argument("--fd-confidence", type=float, default=0.5,
                    help="minimum probability to filter weak detections")
    ap.add_argument("--fd-label-encoder", type=str,
                    default=relPath(["models", "le.pickle"]),
                    help="path to label encoder")
    ap.add_argument("--fd-dataset", type=str,
                    default=relPath(["dataset"]),
                    help="path to input directory of faces + images")
    ap.add_argument("--fd-model", type=str,
                    default=relPath(["models", "openface_nn4.small2.v1.t7"]),
                    help="path to OpenCV's deep learning face embedding model")
    ap.add_argument("--fd-recognizer", type=str,
                    default=relPath(["models", "recognizer.pickle"]),
                    help="path to model trained to recognize faces")
    ap.add_argument("--fd-train", action="store_true",
                    help="train face recognizer on start")
    ap.add_argument("--last-faces", type=int, default=60,
                    help="show detected faces for last seconds")
    ap.add_argument("--mdclip-codec", type=str, default="mp4v", choices=["mp4v", "XVID"],
                    help="video codec")
    ap.add_argument("--mdclip-buf-length", type=int, default=10,
                    help="buffered pre motion detection clip length in seconds")
    ap.add_argument("--mdclip-max-length", type=int, default=300,
                    help="maximal after motion detection clip length in seconds")
    ap.add_argument("--mdclip-min-length", type=int, default=10,
                    help="minimal after motion detection clip length in seconds")
    ap.add_argument("--md-debounce", type=int, default=0.4,
                    help="debounce motion detection in seconds")
    ap.add_argument("--thumb-size", type=int, default=150,
                    help="thumb side size")
    ap.add_argument("--tk-queue-interval", type=int, default=50,
                    help="process tk app queue with interval in milliseconds")
    ap.add_argument("-g", "--geometry", type=str, default="800x640",
                    help="window geometry")
    ap.add_argument("-l", "--loglevel", type=int, default="20", choices=[0, 10, 20, 30, 40, 50],
                    help="log level 0:NOTSET, 10:DEBUG, 20:INFO, 30:WARNING, 40:ERROR, 50:CRITICAL")
    ap.add_argument("-s", "--source", type=str, default=0,
                    help="path to input video, camera otherwise")
    ap.add_argument("-w", "--workdir", type=str,
                    default=relPath(["datastore"]),
                    help="workdir for store images, videos")
    return ap.parse_args()


def relPath(pathItems):
    return os.path.sep.join([os.path.dirname(os.path.realpath(__file__))] + pathItems)


def loopVideo(opts=None, whileFn=lambda: True, getActionLevel=lambda: 0):

    # initialize the video stream, then allow the camera sensor to warm up
    logger.info("starting video stream")
    # vs = cv2.VideoCapture(opts["source"])
    vs = BufferedVideoStream(opts["source"], depth=opts["mdclip_buf_length"]*20)
    time.sleep(2.0)

    clipMaker = ClipMaker(vs, codec=opts["mdclip_codec"], thumbsize=opts["thumb_size"], workdir=opts["workdir"])
    faceR = FaceRecognizer(vs,
                           detector=opts["face_detector"],
                           model=opts["fd_model"],
                           recognizer=opts["fd_recognizer"],
                           labelEncoder=opts["fd_label_encoder"],
                           detectorConfidence=opts["fd_confidence"],
                           getActionLevel=getActionLevel)

    faceRec = FaceRecorder(path=opts["workdir"])
    mdState = MDState(False, None, 0, 0)
    lastFaces = []

    # loop over frames from the video file stream
    while vs.isOpened() and whileFn():
        # define the time mark of current frame
        tmark = time.monotonic()

        # grab the frame from the threaded video stream
        ok, frame = vs.read()
        if not ok:
            continue

        # resize image, then use for motion detection
        frameSized = cv2.resize(frame, (300, 300))
        mdState = detectMotion(mdState, frameSized, tmark=tmark, debounce=opts["md_debounce"])

        # store frame and draw motion icon
        if mdState.detected:
            showMotion()

            if tmark - clipMaker.startedAt >= opts["mdclip_max_length"]:
                clipMaker.stop()

            if not clipMaker.started:
                clipMaker.start()

        elif tmark - mdState.lastDetectedAt > opts["mdclip_min_length"]:
            clipMaker.stop()

        if getActionLevel() > 0:
            frm, cards = faceR.getFrameCards()
            faceRec.storeFrameCards(frm, cards)
            lastFaces[:] = updateLastFaces(lastFaces, cards, outdate=timedelta(seconds=opts["last_faces"]))
            frame = drawRecognizedFaces(frame, cards)
            frame = drawLastFaces(frame, lastFaces, dataset=opts["fd_dataset"])

        # show the output frame
        drawFrame(frame)
        time.sleep(0.02)

    # cleanup
    try:
        clipMaker.stop()
        faceRec.stop()
        faceR.stop()
        vs.release()
    except:
        pass


def drawFrame(frame):
    global tkApp
    try:
        tkApp.drawFrame(frame)
    # RuntimeError: main thread is not in main loop
    except:
        tkApp.ctrlQueue.put({"subj": "drawFrame", "frame": frame})


def showMotion():
    global tkApp
    try:
        tkApp.showMotion()
    # RuntimeError: main thread is not in main loop
    except:
        tkApp.ctrlQueue.put({"subj": "showMotion"})


def showWrite():
    global tkApp
    try:
        tkApp.showWrite()
    # RuntimeError: main thread is not in main loop
    except:
        tkApp.ctrlQueue.put({"subj": "showWrite"})


def drawRecognizedFaces(frame, cards, color=(255, 0, 0)):
    for i, card in enumerate(cards):
        # draw the bounding box of the face
        startX, startY, endX, endY = card.rect
        cv2.rectangle(frame, (startX, startY), (endX, endY), color, 2)
    return frame


def drawLastFaces(frame, cards, dataset="", sort=lambda card: card.time, dims=(120, 120)):
    frameH, frameW = frame.shape[:2]
    iconH, iconW = dims
    wc = frameW // iconW
    for i, card in enumerate(sorted(cards, key=sort)[-wc:]):
        if card.name and card.name != "unknown":
            frame[0:iconH, i*iconW:(i*iconW+iconW)] = getIcon(dataset, card.name, dims)
        else:
            frame[0:iconH, i*iconW:(i*iconW+iconW)] = createIcon(card.img, dims)
        if card.name:
            cv2.putText(frame, card.name, (i*iconH+4, iconW-8),
                        cv2.FONT_HERSHEY_SIMPLEX, 0.8, (0, 191, 255), 2)
    return frame


def updateLastFaces(lastFaces, cards, outdate=timedelta(seconds=60)):
    for card in cards:
        found = False
        for i, f in enumerate(lastFaces):
            if f.name and card.name:
                # find correspond card by name
                if f.name == card.name:
                    lastFaces[i] = card
                    found = True
                    break
            else:
                # find correspond card by area
                if cardsOverlap(f, card):
                    lastFaces[i] = card
                    found = True
                    break
        if not found:
            lastFaces.append(card)
    # remove outdated
    t0 = datetime.now()
    return [card for card in lastFaces if t0 - card.time < outdate]


def createIcon(img, dims=(300, 300)):
    # create gray blank
    blank = np.zeros((dims[0], dims[1], 3), np.uint8)
    blank[:] = (128, 128, 128)
    # place resized icon at centre of the blank
    (icoH, icoW) = img.shape[:2]
    if icoH > icoW:
        ico = imutils.resize(img, height=dims[0])
        (icoH, icoW) = ico.shape[:2]
        startX = int((dims[1]-icoW)/2)
        blank[0:dims[0], startX:startX + icoW] = ico
    else:
        ico = imutils.resize(img, width=dims[1])
        (icoH, icoW) = ico.shape[:2]
        startY = int((dims[0]-icoH)/2)
        blank[startY:startY+icoH, 0:dims[1]] = ico

    return blank


@functools.lru_cache(maxsize=32)
def getIcon(pathDataset, name, dims=(300, 300)):
    fnames = glob.glob(os.path.sep.join([pathDataset] + [name, "icon.[jp][pn][eg]*"]))
    if not fnames:
        fnames = glob.glob(os.path.sep.join([pathDataset] + [name, "*.[jp][pn][eg]*"]))
    if not fnames:
        return None
    # create gray blank
    blank = np.zeros((dims[0], dims[1], 3), np.uint8)
    blank[:] = (128, 128, 128)
    # place resized icon at centre of the blank
    ico = cv2.imread(fnames[0])
    (icoH, icoW) = ico.shape[:2]
    if icoH > icoW:
        ico = imutils.resize(ico, height=dims[0])
        (icoH, icoW) = ico.shape[:2]
        startX = int((dims[1]-icoW)/2)
        blank[0:dims[0], startX:startX + icoW] = ico
    else:
        ico = imutils.resize(ico, width=dims[1])
        (icoH, icoW) = ico.shape[:2]
        startY = int((dims[0]-icoH)/2)
        blank[startY:startY+icoH, 0:dims[1]] = ico

    return blank


def detectMotion(mdState, frame, tmark=0, debounce=0.04):
    if tmark == 0:
        tmark = time.monotonic()
    # debounce motion detection
    if not mdState.detected or (tmark - mdState.trackedAt) > debounce:
        # https://medium.com/@pytholabs/motion-detector-alarmusing-opencv-python-b3ff21c69e0c
        # convert to gray-scale and add gaussian blur to reduce noise
        frame1 = cv2.GaussianBlur(cv2.cvtColor(frame, cv2.COLOR_BGR2GRAY), (21, 21), 0)
        if mdState.trackedFrame is not None:
            frameDelta = cv2.absdiff(frame1, mdState.trackedFrame)
            # apply a threshold function to suppress the smaller values below 25 which will yield a matrix with binary values
            td = cv2.threshold(frameDelta, 25, 255, cv2.THRESH_BINARY)[1]
            if(td.sum() > 100):
                mdState = MDState(True, frame1, tmark, tmark)
            else:
                mdState = MDState(False, frame1, tmark, mdState.lastDetectedAt)
        else:
            mdState = MDState(False, frame1, tmark, mdState.lastDetectedAt)
    return mdState


def main():
    opts = vars(getCLIArgs())
    logging.basicConfig(level=opts["loglevel"])
    logger.info("start with %s", json.dumps(opts, indent=4))

    if opts["fd_train"]:
        logger.info("train face recognizer")
        extractEmbeddings(dataset=opts["fd_dataset"],
                          detector=opts["face_detector"],
                          embeddings=os.path.sep.join([opts["face_detector"], "embeddings.pickle"]),
                          model=opts["fd_model"],
                          detectorConfidence=opts["fd_confidence"])
        trainModel(embeddings=os.path.sep.join([opts["face_detector"], "embeddings.pickle"]),
                   recognizer=opts["fd_recognizer"],
                   labelEncoder=opts["fd_label_encoder"])
    else:
        time.sleep(2.0)

    global tkApp
    tkApp = TkApp(opts=opts)
    tkApp.run()
    videoThread = threading.Thread(target=loopVideo, args=(
        opts,
        lambda: tkApp.running,
        lambda: tkApp.actionLevel if tkApp.tglMonitor else 0
    ))
    videoThread.start()
    tkApp.eventLoop()


if __name__ == "__main__":
    main()
